File: src/prod/src/ServiceModel/ProgressUnitType.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;

namespace ServiceModel
{
    namespace ProgressUnitType
    {
        FABRIC_PROGRESS_UNIT_TYPE ToPublicApi(Enum val)
        {
            switch (val)
            {
            case Bytes:
                return FABRIC_PROGRESS_UNIT_TYPE_BYTES;

            case ServiceSubPackages:
                return FABRIC_PROGRESS_UNIT_TYPE_SERVICE_SUB_PACKAGES;

            case Files:
                return FABRIC_PROGRESS_UNIT_TYPE_FILES;

            default:
                return FABRIC_PROGRESS_UNIT_TYPE_INVALID;
            }
        }

        Enum FromPublicApi(FABRIC_PROGRESS_UNIT_TYPE val)
        {
            switch (val)
            {
            case FABRIC_PROGRESS_UNIT_TYPE_BYTES:
                return Bytes;

            case FABRIC_PROGRESS_UNIT_TYPE_SERVICE_SUB_PACKAGES:
                return ServiceSubPackages;

            case FABRIC_PROGRESS_UNIT_TYPE_FILES:
                return Files;

            default:
                return Invalid;
            }
        }

        void WriteToTextWriter(__in Common::TextWriter & w, Enum val)
        {
            switch (val)
            {
            case Invalid:
                w << L"Invalid";
                return;

            case Bytes:
                w << L"Bytes";
                return;

            case Files:
                w << L"Files";
                return;

            default:
                w << L"ProgressUnitType(" << static_cast<uint>(val) << L")";
            }
        }

        ENUM_STRUCTURED_TRACE( ProgressUnitType, Invalid, LastValidEnum )
    }
}

File: src/prod/src/managed/Api/src/System/Fabric/ProgressUnitType.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric
{
    using System.Fabric.Interop;

    /// <summary>
    /// Describes the unit of measurement for each <see cref="System.Fabric.IImageStoreProgressHandler.UpdateProgress"/> callback.
    /// </summary>
    public enum ProgressUnitType
    {
        /// <summary>
        /// Reserved for internal use. May indicate a mismatch between managed and native DLL versions.
        /// </summary>
        Invalid = NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_INVALID,

        /// <summary>
        /// The unit of measurement is in bytes.
        /// </summary>
        Bytes = NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_BYTES,
            
        /// <summary>
        /// The unit of measurement is in number of code, config, and data packages.
        /// </summary>
        ServiceSubPackages = NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_SERVICE_SUB_PACKAGES,
            
        /// <summary>
        /// The unit of measurement is in number of files.
        /// </summary>
        Files = NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_FILES 
    }

    internal class ProgressUnitTypeHelper
    {
        public static ProgressUnitType FromNative(NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE native)
        {
            switch (native)
            {
                case NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_BYTES:
                case NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_SERVICE_SUB_PACKAGES:
                case NativeImageStore.FABRIC_PROGRESS_UNIT_TYPE.FABRIC_PROGRESS_UNIT_TYPE_FILES:
                    return (ProgressUnitType)native;

                default:
                    return ProgressUnitType.Invalid;
            }
        }
    }
}